pkg/integrations/cadvisor/cadvisor.go

```diff
@@ -0,0 +1,255 @@
+	StorageDuration time.Duration `yaml:"storage_duration,omitempty"`
+
+	// Containerd config options
+	// Containerd containerd endpoint
+	Containerd string `yaml:"containerd,omitempty"`
+
+	// ContainerdNamespace containerd namespace
+	ContainerdNamespace string `yaml:"containerd_namespace,omitempty"`
+
+	// Docker config options
+	// Docker docker endpoint
+	Docker string `yaml:"docker,omitempty"`
+
+	// DockerTLS use TLS to connect to docker
+	DockerTLS bool `yaml:"docker_tls,omitempty"`
+
+	// DockerTLSCert path to client certificate
+	DockerTLSCert string `yaml:"docker_tls_cert,omitempty"`
+
+	// DockerTLSKey path to private key
+	DockerTLSKey string `yaml:"docker_tls_key,omitempty"`
+
+	// DockerTLSCA path to trusted CA
+	DockerTLSCA string `yaml:"docker_tls_ca,omitempty"`
+
+	// Raw config options
+	// DockerOnly only report docker containers in addition to root stats
+	DockerOnly bool `yaml:"docker_only,omitempty"`
+}
+
+// UnmarshalYAML implements yaml.Unmarshaler for Config
+func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultConfig
+
+	type plain Config
+	err := unmarshal((*plain)(c))
+	// Clear default disabled metrics if explicit disabled metrics are configured
+	if len(c.DisabledMetrics) > 0 {
+		c.disabledMetricsSet = container.MetricSet{}
+	}
+	for _, d := range c.DisabledMetrics {
+		if err := c.disabledMetricsSet.Set(d); err != nil {
+			return fmt.Errorf("failed to set disabled metric: %w", err)
+		}
+	}
+
+	for _, e := range c.EnabledMetrics {
+		if err := c.enabledMetricsSet.Set(e); err != nil {
+			return fmt.Errorf("failed to set enabled metric: %w", err)
+		}
+	}
+
+	if len(c.enabledMetricsSet) > 0 {
+		c.includedMetrics = c.enabledMetricsSet
+	} else {
+		c.includedMetrics = container.AllMetrics.Difference(c.disabledMetricsSet)
+	}
+	return err
+}
+
+// Name returns the name of the integration that this config represents.
+func (c *Config) Name() string {
+	return "cadvisor"
+}
+
+// CommonConfig returns the common settings shared across all configs for
+// integrations.
+func (c *Config) CommonConfig() config.Common {
+	return c.Common
+}
+
+// InstanceKey returns the hostname:port of the cadvisor API server.
+func (c *Config) InstanceKey(agentKey string) (string, error) {
+	return agentKey, nil
+}
+
+// NewIntegration creates a new cadvisor integration
+func (c *Config) NewIntegration(logger log.Logger) (integrations.Integration, error) {
+	return New(logger, c)
+}
+
+func init() {
+	integrations.RegisterIntegration(&Config{})
+}
+
+// New creates a new cadvisor integration
+func New(logger log.Logger, c *Config) (integrations.Integration, error) {
+	klog.SetLogger(logger)
+
+	// Do gross global configs. This works, so long as there is only one instance of the cAdvisor integration
+	// per host.
+	// Containerd
+	containerd.ArgContainerdEndpoint = &c.Containerd
+	containerd.ArgContainerdNamespace = &c.ContainerdNamespace
```

Comment:
So, these are why this is a singleton integration.

I could probably break it down and instantiate the collectors directly with a bit more effort, but this feels "okay" as an MVP.

Reply:
nit: All of the globals being set here should instead be set in the integration's `Run` method the integration instead of New. 

`Run` is guaranteed to never be invoked concurrently for the same integration. However, `New` _may_ be called while `Run` is being invoked, which would potentially lead to a race condition of these globals here. 

Setting them in `Run` will avoid that potential race condition. Changing where these are being initialized might unfortunately push down a lot of logic to Run, though.

Reply:
All good. I *did* have to offload all of the logic to `Run` but it's okay, because it forced me to go back and realize I hadn't stopped the resource manager, which I am now doing after the context ends. 😎